src/Lecture2.hs

```diff
@@ -268,7 +347,19 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit a) = Right a
+eval vars (Var v) = case lookup v vars of
+                         Nothing -> Left (VariableNotFound v)
+                         Just x  -> Right x
+eval vars (Add exp1 exp2) = let
+                              f    = eval vars
+                              val1 = f exp1
+                              val2 = f exp2
+                             in
+                               case (val1, val2) of
+                                    (Left e, _) -> Left e
+                                    (_, Left e) -> Left e
+                                    (Right x, Right y) -> Right (x+y)
```

Comment:
That's a great pattern matching! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️